solidity/dashboard/src/services/token-staking.service.js

```diff
@@ -1,77 +1,75 @@
 import { contractService } from "./contracts.service"
 import {
-  TOKEN_STAKING_CONTRACT_NAME,
-  TOKEN_GRANT_CONTRACT_NAME,
+  TOKEN_STAKING_CONTRACT_NAME
 } from "../constants/constants"
 import moment from "moment"
 import {
   isCodeValid,
   createManagedGrantContractInstance,
   CONTRACT_DEPLOY_BLOCK_NUMBER,
   ContractsLoaded,
+  Web3Loaded,
 } from "../contracts"
 import { isSameEthAddress } from "../utils/general.utils"
 import { isEmptyArray } from "../utils/array.utils"
+import { getEventsFromTransaction } from "../utils/ethereum.utils"
 
-const fetchDelegatedTokensData = async (web3Context) => {
-  const { yourAddress, grantContract, eth, web3 } = web3Context
-  const [
-    stakedBalance,
-    ownerAddress,
-    beneficiaryAddress,
-    authorizerAddress,
-    initializationPeriod,
-  ] = await Promise.all([
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "balanceOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "ownerOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "beneficiaryOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "authorizerOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "initializationPeriod"
-    ),
+const fetchDelegatedTokensData = async () => {
+  const contractsLoaded = await ContractsLoaded
+  const { grantContract, stakingContract } = contractsLoaded
+  const web3 = await Web3Loaded
+  const { eth } = web3
+  const { defaultAccount: yourAddress } = eth
+
+  let operatorStakedEvents
+  try {
+    operatorStakedEvents = await stakingContract.getPastEvents(
+      "OperatorStaked",
+      {
+        fromBlock: CONTRACT_DEPLOY_BLOCK_NUMBER.stakingContract,
+        filter: { operator: yourAddress },
+      }
+    )
+  } catch (err) {}
+
+  let [stakedBalance, ownerAddress, initializationPeriod] = await Promise.all([
+    stakingContract.methods.balanceOf(yourAddress).call(),
+    stakingContract.methods.ownerOf(yourAddress).call(),
+    stakingContract.methods.initializationPeriod().call(),
   ])
 
-  let isUndelegationFromGrant = true
-  let grantStakeDetails
+  let transactionHash
+  let beneficiaryAddress = ownerAddress
+  let authorizerAddress = ownerAddress
+
+  if (operatorStakedEvents.length > 0) {
+    ;({
+      transactionHash,
+      returnValues: {
+        beneficiary: beneficiaryAddress,
+        authorizer: authorizerAddress,
+      },
+    } = operatorStakedEvents[0])
```

Comment:
Thanks for the input :) 

I actually used [destructurization without declaration](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#:~:text=Assignment%20without%20declaration,destructuring%20assignment%20without%20a%20declaration.) here, which in my opinion is better because we don't have to declare new variables (like transactionHash, beneficiary and authorizer in your example) just to assign them to the new ones. In my example we destructurize operatorStakedEvents[0] and at the same time assign them to already declared, better-named variables.

The weird look is probably due to the semicolon at the beginning, which is actually [needed when writing JavaScript without semicolons](https://stackoverflow.com/a/56068605) which i believe we do.

I do agree however that it might be less readable, but still this solution does not require to declare new variables. Maybe a better solution would be to write a comment that describes what happens here?

Reply:
yeah I see your point.

> Maybe a better solution would be to write a comment that describes what happens here?

I do agree 👍

Reply:
I'm going to be annoying here and agree with Rafal's original point, and disagree that a comment is a better solution here (even though we use this pattern in tests in a few cases where it's hard to avoid). I think being explicit about the reassignment rather than using destructuring is clearer in cases where we start needing non-obvious syntax noise to achieve what we're trying to achieve.

Functionally speaking, I consider the fact that we're trying to reassign to variables as a _hint_ (though not a mandate) that we should consider breaking this functionality out into its own function. Combined with Rafal's feedback further up about checking for events before doing the balance/owner/etc lookups, perhaps that function can be something like `delegationInfoFromStakedEvents` and wrap all that logic?

Without going through a whole refactor, another pattern we might use here is something like:

```javascript
  const {
    transactionHash,
    returnValues: {
      beneficiary: beneficiaryAddress,
      authorizer: authorizerAddress
    }
  } = operatorStakedEvents[0] || {
    // Defaults.
    returnValues: { beneficiary: ownerAddress, authorizer: ownerAddress }
  }
```

I consider this clearer if we start using it consistently.

Reply:
Thanks, this seems like a reasonable solution 😄 👍 I will implement this.

Reply:
I wrapped all the logic you mentioned in `delegationInfoFromStakedEvents` function